feat: push docker image to ghcr.io

ayewo opened the pull request:
**Summary:**  
_Build and push a Docker image to GHCR.io._

**Issue Reference(s):**  
Fixes #502. Please also see https://github.com/tailcallhq/tailcallhq.github.io/pull/22 to update the [docs](https://tailcall.run/docs/guides/installation/#docker) where the Docker image is used.

When you push a new tag, say v0.13.0, the action will apply the tags `v0.13.0` and `latest` to it. Please refer to the [documentation](https://github.com/docker/metadata-action#basic).

### Below are examples from my account 
I tested using my account so it is published at`ayewo/tailcall/tc-server:v0.13.0` but once merged, it will be published at `tailcallhq/tailcall/tc-server:v0.13.0`.
The published package:
* https://github.com/ayewo/tailcall/pkgs/container/tailcall%2Ftc-server/138387740?tag=v0.13.0
The successful job run:
* https://github.com/ayewo/tailcall/actions/runs/6552618536/job/17796304111
 

**Build & Testing:**

- [x] I ran `cargo test` successfully.
- [x] I have run `./lint.sh` to address and fix linting issues.

**Checklist:**

- [x] I have added relevant unit & integration tests.
- [x] I have updated the documentation accordingly (if applicable).
- [x] I have performed a self-review of my own code.

/claim #502

ayewo commented:
@tusharmath you'll need to grant the GH Action `write` permissions so it can publish packages at: https://github.com/orgs/tailcallhq/packages

<img width="1664" alt="image" src="https://github.com/tailcallhq/tailcall/assets/20957603/7305d515-04db-4a8d-98e8-c8d38cd06b18">

tusharmath commented on .github/workflows/ghcr.yml:
```diff
@@ -0,0 +1,53 @@
+        with:
+          registry: ${{ env.REGISTRY }}
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
```
```suggestion
          password: ${{ secrets.GHCR_TOKEN }}
```

tusharmath replied:
```suggestion
          password: ${{ secrets.GITHUBTOKEN }}
```

tusharmath commented on .github/workflows/ghcr.yml:
```diff
@@ -0,0 +1,53 @@
+        with:
+          registry: ${{ env.REGISTRY }}
+          username: ${{ github.actor }}
+          password: ${{ secrets.GHCR_TOKEN }}
```
```suggestion
          password: ${{ secrets.GITHUB_TOKEN }}
```

tusharmath commented:
How is this going to work without a dockerfile?

ayewo commented:
> How is this going to work without a dockerfile?

@tusharmath it works just fine. The GitHub Action uses the [Dockerfile](https://github.com/tailcallhq/tailcall/blob/main/Dockerfile) in the root of the repo.

ayewo commented:
@tusharmath if you grant the `GHCR_TOKEN` you added with `package:write` permissions then the GH Action will be able to publish Docker images to GitHub Packages at: https://github.com/orgs/tailcallhq/packages

ayewo commented:
> How is this going to work without a dockerfile?

@tusharmath I created a brand new GitHub organization https://github.com/orgs/tailcall-demo/ to demonstrate that the GH Action I added works as intended. You might have to merge my changes for it to work under the `tailcallhq` organization.

I've added you to it as an Admin. Please visit https://github.com/tailcall-demo to accept the invitation.

1. Docker image built just now using `ghcr.yml`: https://github.com/orgs/tailcall-demo/packages
2. Repo settings page from which I took the screenshot [linked](https://github.com/tailcallhq/tailcall/pull/509#issuecomment-1767220573) above https://github.com/orgs/tailcall-demo/packages/container/tailcall%2Ftc-server/settings
